### ISSUE TITLE
Regent index launch failure with cross_product

### ISSUE THREAD

**Issue by syamajala** [StanfordLegion/legion#1285]
I am seeing the following error when trying to index launch a task over a cross_product partition:

```
...j/legion_s3d_test/legion/language/terra/src/terralib.lua:1202: Errors reported during typechecking.
/lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/codegen.t:2077: invalid conversion from int3d to int64
    [ref_expr.value] = [value_expr.value]
                                 ^
/lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/codegen.t:2077: user-defined cast failed: ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2390: assertion failed!
stack traceback:
        [C]: in function 'assert'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2390: in function <...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:2365>
        [C]: in function 'xpcall'
        /lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/codegen.t:2077: Errors reported during invoking __cast
    [ref_expr.value] = [value_expr.value]
                                 ^
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:2290: in function 'insertcast'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:2665: in function 'trylist'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:2674: in function 'insertcasts'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3212: in function 'checksingle'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3344: in function 'checkstmts'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3082: in function 'docheck'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3120: in function 'checkexp'
        ...
        src/terralist.lua:150: in function 'block'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:8454: in function 'fn'
        src/terralist.lua:150: in function 'block'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:11159: in function 'compile_thunk'
        ...aj/legion_s3d_test/legion/language/src/regent/std_base.t:1616: in function 'complete'
        ...aj/legion_s3d_test/legion/language/src/regent/std_base.t:1077: in function 'get_definition'
        ...aj/legion_s3d_test/legion/language/src/regent/std_base.t:1172: in function 'make_wrapper'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4332: in function 'make_task_wrappers'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4888: in function 'save_tasks'
        s3d.rg:2113: in main chunk
    [ref_expr.value] = [value_expr.value]
                                 ^

stack traceback:
        [C]: in function 'error'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:396: in function 'finishandabortiferrors'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:3437: in function 'typecheck'
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:1202: in function 'definequote'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:2073: in function 'write'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:9880: in function 'fn'
        src/terralist.lua:150: in function 'block'
        /lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/codegen.t:8740: in function 'userfn'
        /lustre/scratch/vsyamaj/legion_s3d_test/legion/language/src/regent/codegen.t:8740: Errors reported during evaluating Lua code from Terra
      [cleanup_after(cx, codegen.block(cx, block))]
                   ^
        ...j/legion_s3d_test/legion/language/terra/src/terralib.lua:1843: in function 'evalluaexpression'
        ...
        src/terralist.lua:150: in function 'block'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:8454: in function 'fn'
        src/terralist.lua:150: in function 'block'
        ...maj/legion_s3d_test/legion/language/src/regent/codegen.t:11159: in function 'compile_thunk'
        ...aj/legion_s3d_test/legion/language/src/regent/std_base.t:1616: in function 'complete'
        ...aj/legion_s3d_test/legion/language/src/regent/std_base.t:1077: in function 'get_definition'
        ...aj/legion_s3d_test/legion/language/src/regent/std_base.t:1172: in function 'make_wrapper'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4332: in function 'make_task_wrappers'
        ...vsyamaj/legion_s3d_test/legion/language/src/regent/std.t:4888: in function 'save_tasks'
        s3d.rg:2113: in main chunk
make[1]: *** [Makefile:266: /lustre/scratch/vsyamaj/legion_s3d_test/build/hept/libregent_tasks.so] Error 1
make[1]: Leaving directory '/lustre/scratch/vsyamaj/legion_s3d_test/rhst'
make: *** [Makefile:82: rhst_hept] Error 2
```

I have not been able to find a reproducer yet as standalone examples outside of S3D seem to be working.

**Comment by syamajala** [comment]
The launch looks like this:

```
    __demand(__index_launch)
    for color in is_rank_subrank do
      CalcVolumeTask(lp_q_subrank[int3d{color.x, color.y, color.z}][color.w],
                     lp_int_subrank[int3d{color.x, color.y, color.z}][color.w])
    end
```

and works if I change it to `__forbid(__index_launch)`

**Comment by rupanshusoi** [comment]
@elliottslaughter I think my code is at fault here, but it's hard to tell exactly where without a reproducer.

**Comment by syamajala** [comment]
There was another bug that was fixed recently that has a reproducer: https://github.com/StanfordLegion/legion/issues/1280

I tried modifying that reproducer to see if I get it to hit this error, but haven't had any luck.

**Comment by elliottslaughter** [comment]
The error goes away with `-findex-launch-dynamic 0 -foverride-demand-index-launch 1`. This seems to indicate the issue is in dynamic checks. But I don't have a reproducer yet so it's hard to tell exactly where or how it's occurring.

**Comment by elliottslaughter** [comment]
This should be fixed in `master` with 6040eb65aac99d78f2a848384f580e59d5199f5c. However note that even with the fix, I think our current dynamic check implementation is not sufficient for this case and will end up generating errors at runtime, so you still need the flags I posted in my last comment.

**Comment by syamajala** [comment]
This is fixed.
